dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/compute/strategy/GetOrCreateNetworkDomainThenCreateNodes.java

```diff
@@ -0,0 +1,175 @@
+      if (networkDomainId != null) {
+         logger.debug("Found a suitable existing network domain %s", networkDomainId);
+      } else {
+         networkDomainId = deployNeworkDomain(datacenterId, networkDomainName);
+      }
+      return networkDomainId;
+   }
+
+   private String deployNeworkDomain(final String datacenter, final String networkDomainName) {
+      logger.debug("Creating a network domain '%s' in Datacenter '%s' ...", networkDomainName, datacenter);
+      String networkDomainId = api.getNetworkApi()
+            .deployNetworkDomain(datacenter, networkDomainName, "network domain created by jclouds",
+                  NetworkDomain.Type.ESSENTIALS.name());
+      String message = format("networkDomain(%s) is not ready within %d ms.", networkDomainId, timeouts.nodeRunning);
+      boolean isNetworkDomainReady = retry(new NetworkDomainState(api.getNetworkApi(), State.NORMAL),
+            timeouts.nodeRunning).apply(networkDomainId);
+      if (!isNetworkDomainReady) {
+         throw new IllegalStateException(message);
+      }
+      return networkDomainId;
+   }
+
+   private Optional<String> getExistingNetworkDomainId(final String datacenterId, final String networkDomainName) {
+      Optional<NetworkDomain> networkDomainOptional = api.getNetworkApi()
+            .listNetworkDomainsWithDatacenterIdAndName(datacenterId, networkDomainName).concat().first();
+      if (networkDomainOptional.isPresent()) {
+         return Optional.of(networkDomainOptional.get().id());
+      } else {
+         return Optional.<String>absent();
+      }
+   }
+
+   private String tryCreateOrGetExistingVlanId(final String networkDomainId, final String vlanName,
+         final DimensionDataCloudControlTemplateOptions templateOptions) {
+      String vlanId = getExistingVlan(networkDomainId, vlanName).orNull();
+      if (vlanId != null) {
+         logger.debug("Found a suitable existing vlan %s", vlanId);
+      } else {
+         vlanId = deployVlan(networkDomainId, vlanName, templateOptions);
+      }
+      return vlanId;
+   }
+
+   private String deployVlan(final String networkDomainId, final String vlanName,
+         final DimensionDataCloudControlTemplateOptions templateOptions) {
+      logger.debug("Creating a vlan %s in network domain '%s' ...", vlanName, networkDomainId);
+      String defaultPrivateIPv4BaseAddress = firstNonNull(templateOptions.getDefaultPrivateIPv4BaseAddress(),
+            DEFAULT_PRIVATE_IPV4_BASE_ADDRESS);
+      Integer defaultPrivateIPv4PrefixSize = firstNonNull(templateOptions.getDefaultPrivateIPv4PrefixSize(),
+            DEFAULT_PRIVATE_IPV4_PREFIX_SIZE);
+
+      String vlanId = api.getNetworkApi()
+            .deployVlan(networkDomainId, vlanName, "vlan created by jclouds", defaultPrivateIPv4BaseAddress,
+                  defaultPrivateIPv4PrefixSize);
+      boolean isVlanReady = retry(new VlanState(api.getNetworkApi(), State.NORMAL), timeouts.nodeRunning).apply(vlanId);
```

Comment:
This is a small PR and I think we should introduce the Guice-configured predicates here and stop initializing them manually. Let's better properly configure now the predicates now that we have the compute service module and get them injected where needed. This can be done in a different PR that unifies them and removes the static utils class.